pennylane/qnodes/base.py

```diff
@@ -482,29 +496,21 @@ def _construct(self, args, kwargs):
         self.obs_queue = []  #: list[Observable]: applied observables
 
         # set up the context for Operator entry
-        if qml._current_context is None:
-            qml._current_context = self
-        else:
-            raise QuantumFunctionError(
-                "qml._current_context must not be modified outside this method."
-            )
-        try:
-            # generate the program queue by executing the quantum circuit function
-            if self.mutable:
-                # it's ok to directly pass auxiliary arguments since the circuit is re-constructed each time
-                # (positional args must be replaced because parameter-shift differentiation requires VariableRefs)
-                res = self.func(*self.arg_vars, **kwargs)
-            else:
-                # TODO: Maybe we should only convert the kwarg_vars that were actually given
-                res = self.func(*self.arg_vars, **self.kwarg_vars)
-        except:
-            # The qfunc call may have failed because the user supplied bad parameters, which is why we must wipe the created VariableRefs.
-            self.arg_vars = None
-            self.kwarg_vars = None
-
-            raise
-        finally:
-            qml._current_context = None
+        with self:
+            try:
+                # generate the program queue by executing the quantum circuit function
+                if self.mutable:
+                    # it's ok to directly pass auxiliary arguments since the circuit is re-constructed each time
+                    # (positional args must be replaced because parameter-shift differentiation requires VariableRefs)
+                    res = self.func(*self.arg_vars, **kwargs)
+                else:
+                    # TODO: Maybe we should only convert the kwarg_vars that were actually given
+                    res = self.func(*self.arg_vars, **self.kwarg_vars)
+            except:
+                # The qfunc call may have failed because the user supplied bad parameters, which is why we must wipe the created VariableRefs.
+                self.arg_vars = None
+                self.kwarg_vars = None
+                raise
```

Comment:
This is much cleaner! In terms of logical flow, however, nothing should have changed as far as I can tell? The previous use of try-except-finally should be replicating what the context manager is doing in the new version (i.e., an exception causes the global context variable to be reset safely)

Reply:
Yes, and no, nothing should have changed in terms of the flow. This PR would also be a part of what's required to implement @johannesjmeyer 's queuing ADR.

Reply:
:+1: